File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char,int> m;

        for(int i=0;i<s.size();i++)
        {
            m[s[i]]++;
        }

        string ans = "";
        priority_queue<pair<int,char>> maxHeap;

        for(auto it:m)
        {
            maxHeap.push({it.second, it.first});
        }
        
        while(!maxHeap.empty())
        {
            int count = maxHeap.top().first;
            char ch = maxHeap.top().second;
            maxHeap.pop();
            ans+=string(count, ch);
        }
        return ans;
    }
};

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {

        HashMap<Character, Integer> mpp = new HashMap<>();

        for (char ch : s.toCharArray()) {
            mpp.put(ch, mpp.getOrDefault(ch, 0) + 1);
        }

        PriorityQueue<Character> pq = new PriorityQueue<>(
                (a, b) -> mpp.get(b) - mpp.get(a));
        for (char ch : mpp.keySet()) {
            pq.add(ch);
        }
        StringBuilder ans = new StringBuilder();
        while(!pq.isEmpty()) {
            char ch = pq.poll();
            int freq = mpp.get(ch);

            while(freq > 0)
            {
                ans.append(ch);
                freq--;
            }
            
        }

        return ans.toString();

    }
}